Feature/logbook structured rpc logging (#1703, #1704)

beckjake opened the pull request:
Fixes #1703 
Fixes #1704 

As part of this PR, I stumbled on a pair of odd connection related bugs that I fixed.

- Log compilation to an array of messages as well as stdout/`dbt.log`, and return that in the `status` endpoint. (#1703)
- Log using the `logbook` package, spit out json in `dbt rpc` (#1704)
  - This json output matches the data returned in the `logs` field of rpc responses
- `dbt.logger` now provides a `log_manager` that lets you toggle the various global logging settings: 
  - debug mode (level + text format changes for `dbt --debug ...`)
  - human-readable text output vs json (for `dbt rpc`)
  - stdout vs stderr (for `dbt ls`)
- made `dbt/rpc.py` into a package `dbt/rpc`, added it to `mypy.ini`.
- Added some convenience methods to the log manager for logging control

This will break anything that hooks in to the dbt logger :)

drewbanin commented on core/dbt/rpc/response_manager.py:
```diff
@@ -0,0 +1,275 @@
+            # calling close(), the connection in the parent ends up throwing
+            # 'connection already closed' exceptions
+            if os.name != 'nt':
+                cleanup_connections()
```
I'm still digging into this, but when we cleanup connections here, will that discard the relation cache?

beckjake replied:
It shouldn't, it should only remove the actual connection handles.

drewbanin commented on core/dbt/rpc/task_manager.py:
```diff
@@ -0,0 +1,196 @@
+        return result
+
+    def process_kill(self, task_id):
+        # TODO: this result design is terrible
```
is this todo for this pr, or a future one?

beckjake replied:
This is an old comment that I copied over, but it's still true - so a future PR.

drewbanin commented on core/dbt/adapters/sql/connections.py:
```diff
@@ -50,16 +53,25 @@ def add_query(self, sql, auto_begin=True, bindings=None,
+                log_sql = sql
+
+            logger.debug(
+                'On {connection_name}: {sql}....',
```
this is super minor, but we previously only appended ellipses if the sql was abridged. Can we preserve that behavior here? The benefit is just that the logged sql can be copy/pasted as-is from the logs

beckjake commented on core/dbt/logger.py:
```diff
@@ -46,170 +34,381 @@ def notice(self, msg, *args, **kwargs):
 
-initialized = False
+STDOUT_LOG_FORMAT = '{record.message}'
+# TODO: can we change the time to just "{record.time:%Y-%m-%d %H:%M:%S.%f%z}"?
```
The first line is really complicated and silly looking (`'{record.time:%Y-%m-%d %H:%M:%S%z},{record.time.microsecond:03}'` -> `2019-08-30 11:54:24,853`) to try and exactly copy logging's default pythong `logging` format. I'd prefer the logbook default time format (`'{record.time:%Y-%m-%d %H:%M:%S.f%z}` -> `2019-08-30 11:54:24.853978`), as it's both more precise and not comma-separated. But changing it will break anyone using regex on dbt debug output, so I wanted to leave it alone to start.

drewbanin replied:
that makes a lot of sense! Let's leave it as is for now.

If we did want to break this, now would be the time to do it though!!

beckjake commented on core/dbt/main.py:
```diff
@@ -125,25 +124,31 @@ def initialize_config_values(parsed):
 
-    if parsed.record_timing_info:
-        profiler_enabled = True
+        # we've parsed the logs - we can now decide if we're debug or not
```
Yes!

beckjake commented:
>You mentioned that you "stumbled on a pair of odd connection related bugs that I fixed" - can you say more about that?

The first I fixed in [this commit](https://github.com/fishtown-analytics/dbt/pull/1715/commits/c621e7d38) - on failed connections, we would have `connection.handle = None` (because we got no handle) but were still calling `close()` on the handle from `SQLConnectionManager.cancel_open`

The second was in the integration tests - in some failure paths, we did a similar thing where we tried to rollback a handle, but the handle might be None, which obscured the actual failing error message.
